### PR TITLE
#7530: Improve FD2 runtime arg performance

### PR DESCRIPTION
**Pull request by pgkeller**

### CODE REVIEW

**Review by kmabeeTT — APPROVED** [review]
lgtm!

File: tt_metal/impl/program/program.cpp [hunk 1]
```diff
@@ -835,40 +832,63 @@ void Program::update_runtime_args_transfer_info(Device* device) {
...
+            auto kernel = detail::GetKernel(*this, kernel_id);
 
-        // Common Runtime Args (Multicast)
-        const auto &common_rt_args = kernel->common_runtime_args();
+            uint32_t dst = processor_to_l1_arg_base_addr.at(kernel->processor());
 
-        if (common_rt_args.size() > 0) {
-            uint32_t common_args_addr = dst + kernel->get_common_runtime_args_offset();
-            vector<pair<uint32_t, uint32_t>> dst_noc_multicast_info =
-                extract_dst_noc_multicast_info<std::vector<CoreRange>>(device, kernel->logical_coreranges(),kernel->get_kernel_core_type());
```
Comment (aliuTT) [inline tt_metal/impl/program/program.cpp:865]: Not applicable to the refactor here, but are Common Runtime Args always same as kernel logical core range?

**Comment by kmabeeTT** [comment]
For Allan Q, I'm not sure if email reply here works (let's see) but yes
they are the same.

File: tt_metal/impl/program/program.cpp (hunk 1 above)
Comment (kmabeeTT) [reply tt_metal/impl/program/program.cpp:865]: Yes this is true, common runtime args are always applied to kernel->logical_coreranges()

File: tt_metal/impl/program/program.hpp [hunk 2]
```diff
@@ -133,6 +133,7 @@ class Program {
     // The buffer that holds the kernel/binaries/etc for this program
     std::vector<std::unique_ptr<Buffer>> kg_buffers;
     std::unique_ptr<Buffer> buffer;
+    bool runtime_args_dirty;
```
Comment (abhullar-tt) [inline tt_metal/impl/program/program.hpp:136]: what sets this when runtime args are updated by user app?

Comment (aliuTT) [reply tt_metal/impl/program/program.hpp:136]: I think the idea is that when runtime args are updated, we can reuse the transfer_infos that are populated when the program was created with `runtime_args_dirty=true`.

Comment (pgkeller) [reply tt_metal/impl/program/program.hpp:136]: yes, really there should be 2 levels of dirty. right now we'll always re-populate the data but we don't rebuild the structure.

**Review by abhullar-tt — APPROVED** [review]

**Pull request merged**